mocha-shared-behaviors/mocha-shared-behaviors.md

```diff
@@ -0,0 +1,471 @@
+>   refactoring than if you had just listed out your tests separately.
+
+So writing shared behaviors using the "`this` keyword" does work. And it can be pretty useful from time to time.
+But it can also bring a lot of unneeded complexity to your tests.
+
+**Avoid using the Mocha context as much as you can!**
+_Same thing for shared behaviors in general!_
+
+Let's deconstruct the previous example, and minimize its complexity step-by-step.
+
+## integrated setup or tear down
+
+Back to the ["functional"
+interface](https://github.com/mochajs/mocha/pull/3399) PR. Why would we need a "functional" interface in Mocha in the first place?
+
+Let's try to rewrite the previous example using an arrow function. Of course, a lambda doesn't have a "this", so here
+I'll use its closure.
+
+```javascript
+/// helpers.js
+export function shouldBehaveLikeAUser(user) {
+  it('should have .name.first', () => {
+    expect(user.name.first).to.equal('tobi');
+  });
+  // other tests
+}
+
+/// user.test.js
+describe('User', () => {
+  let user;
+
+  beforeEach(() => {
+    user = new User('tobi', 'holowaychuk');
+  });
+
+  shouldBehaveLikeAUser(user);
+});
+```
+
+Let's run this and...:boom: it fails!
+
+```
+TypeError: Cannot read property 'name' of undefined
+  at Context.name (test/helpers.js:5:17)
+```
+
+This is because Mocha identifies and "records" your test suite first, and _then_ runs your callbacks. So here, it runs
+`beforeEach` and `shouldBehaveLikeAUser` (`user` being undefined at this point) and only _then_ `beforeEach.fn` and `it.fn`.
+
+## "All-in-one"
+
+One solution is to move the `beforeEach` in `shouldBehaveLikeAUser`.
+
+```javascript
+/// helpers.js
+export function shouldBehaveLikeAUser(buildUserFn, { firstName, lastName, fullname }) {
+  let userLike;
+
+  beforeEach(() => {
+    userLike = buildUserFn();
+  });
+  
+  it('should have .name.first', () => {
+    expect(userLike.name.first).to.equal(firstName);
+  });
+  // other tests
+};
+
+/// user.test.js
+describe('User', () => {
+  shouldBehaveLikeAUser(() => new User("tobi", "holowaychuk"), {
+    firstName: "tobi",
+    lastName: "holowaychuk",
+    fullname: 'tobi holowaychuk'
+  });
+});
+
+/// admin.test.js
+describe('Admin', () => {
+  shouldBehaveLikeAUser(() => new Admin("tobi", "holowaychuk"), {
+    firstName: "tobi",
+    lastName: "holowaychuk",
+    fullname: 'tobi holowaychuk'
+  });
+});
+```
+
+Here, nothing is "hidden." Just by looking at the signature, we understand that `shouldBehaveLikeAUser` will test that the constructor you gave will fit the "User" behavior definition. This can be enhanced by adding a JSDoc @param or some TypeScript.
+
+And it's self-sufficient. No side effects or closure requirements here.
+
+**But**, it has its own logic. For example, if you add the first argument to the `Admin` constructor, it fails, because you introduced a bug in the shared behavior, and there is a chance you won't see it directly because you didn't define a spec for that.
```

Comment:
I don't follow this? that is the same for all the solutions right?

if you change
```
() => new Admin("tobi", "holowaychuk")
// to
() => new Admin("extra", "tobi", "holowaychuk")
```

then tests will always fail? e.g. order of parameters is always important and results always in a breaking change

I would not see this as a problem of this particular solution

Reply:
Oh right, it's outdated! This explanation is based on the previous version which used the "construtor" argument. I'll rewrite it. Thx

Reply:
Done! Let me know what you think of this new transition.